server/src/main/java/io/mantisrx/server/master/SchedulingService.java

```diff
@@ -330,28 +333,63 @@ private TaskSchedulingService setupTaskSchedulingService(TaskScheduler taskSched
         return hasValue ? Optional.of(lease.getAttributeMap().get(attributeName).getText().getValue()) : Optional.empty();
     }
 
+    /**
+     * Attempts to launch tasks given some number of leases from Mesos.
+     *
+     * When a task is launched successfully, the following will happen:
+     *
+     * 1. Emit a {@link WorkerLaunched} event to be handled by the corresponding actor.
+     * 2. Makes a call to the underlying Mesos driver to launch the task.
+     *
+     * A task can fail to launch if:
+     *
+     * 1. It doesn't receive enough metadata for {@link WorkerPorts} to pass its preconditions.
+     *      - No launch task request will be made for this assignment result.
+     * 2. It fails to emit a {@link WorkerLaunched} event.
+     *      - The worker will get unscheduled for this launch task request.
+     * 3. There are no launch tasks for this assignment result.
+     *      - All of these leases are rejected.
+     *      - Eventually, the underlying Mesos driver will decline offers since there are no launch task requests.
+     *
+     * @param requests collection of assignment results received by the scheduler.
+     * @param leases list of resource offers from Mesos.
+     */
     private void launchTasks(Collection<TaskAssignmentResult> requests, List<VirtualMachineLease> leases) {
         List<LaunchTaskRequest> launchTaskRequests = new ArrayList<>();
+
         for (TaskAssignmentResult assignmentResult : requests) {
             ScheduleRequest request = (ScheduleRequest) assignmentResult.getRequest();
-            WorkerPorts workerPorts = new WorkerPorts(assignmentResult.getAssignedPorts());
-            boolean success = jobMessageRouter.routeWorkerEvent(new WorkerLaunched(
-                    request.getWorkerId(),
-                    request.getStageNum(),
-                    leases.get(0).hostname(),
-                    leases.get(0).getVMID(),
-                    getAttribute(leases.get(0), slaveClusterAttributeName),
-                    workerPorts));
-            if (success) {
-                launchTaskRequests.add(new LaunchTaskRequest(request, workerPorts));
-            } else {
-                unscheduleWorker(request.getWorkerId(), Optional.ofNullable(leases.get(0).hostname()));
+
+            WorkerPorts workerPorts = null;
```

Comment:
Need a closer review here. Intent is to detect bad worker ports when launching the task. If bad, then don't even send a `WorkerLaunched` event. Expect this worker to be stuck in a state to be resubmitted in the normal stuck/resubmit workflow.

Reply:
think we could proactively unschedule the worker here if the assignmentResult doesn't return the expected number of ports.. Relying on standard flow for resubmit would wait for heartbeat timeouts which could take a while

Reply:
Makes sense. I'll proactively unschedule if there's no worker ports.